Add tag to favorites

TBog opened the pull request:
closes #1118
* long-touching the tag in the tags menu will add it to the favorites
* from the preferences one can add and remove tags from the favorites bar

TODO:
- [x] find a better way to add tags to the favorites
- [x] show tags in the favorites bar
- [ ] test with tags in other languages 
- [x] test tags starting with unicode emoticons

TBog commented:
![device-2019-01-30-130908](https://user-images.githubusercontent.com/6465804/51977568-72b70b80-2490-11e9-857d-388d2476cecc.jpg)

Neamar commented:
Hallo!

Not sure how I feel about the feature, will think some more about it.

Regarding the design... I see you made the favorite a ViewGroup. That's fine, but it adds more views to the screen, which is something we should try to minimize.
Instead, what about generating a drawable that contains the first letter from the tag? That way you can treat tags as a standard result, display their drawable directly, and avoid the need for a viewgroup.
You can even tint the drawable with the main color to make it look nice :D

TBog commented:
- Removed ViewGroup change
- Added white round rectangle background
- Text color changed to see-through
Preview:
![device-2019-03-04-132132](https://user-images.githubusercontent.com/6465804/53737180-76292280-3e94-11e9-9233-dc23ae0dd988.jpg)

Neamar commented:
Looks better :) 

The letter seems to be off-center, can you have a quick look at that?

I'm curious how we can make this feature more discoverable. It can be really powerful, and to be honest even though I personally don't feel I need it it's been asked a lot of times!

In addition to long clicking in the tag bar, maybe add it as a settings in the "Favorites" preference screen?

Neamar commented:
(kinda like what you already did for the tag bar with "Tags to show in menu" -- a similar "Tags to show in favorites"?)

TBog commented:
> (kinda like what you already did for the tag bar with "Tags to show in menu" -- a similar "Tags to show in favorites"?)

Not sure if I got the feature description right ... the fav bar has other stuff also, one would need to select the tags to show and then go to the fav bar and reorder them; is this acceptable?

The letters are centered by StaticLayout

TBog commented:
1. Centered the letters using the rect given by Paint
2. There is some commented code that draws a rounded rectangle around the letter but I don't like it, if someone else has other views on the matter it may be activated by a preference I suppose.
3. Added the "Tags to show in favorites" preference. Not sure if it's the cleanest implementation but it works.

PS. I don't like merges, I only re-base, that's why I force-pushed.

TBog commented:
I ended up not using StaticLayout and also fixed the emoticons issue.
I still think that using some fonts may look bad on the fav bar but I can be fixed when it's reported.
I will remove the WIP from this PR

Neamar commented:
Hey @TBog! There is still one unchecked item on the PR, but do you want me to try to merge it? In terms of feature, it seems to be in a pretty good state right now

TBog commented:
If you are satisfied with the looks of it, you can merge.

Neamar commented on app/src/main/java/fr/neamar/kiss/dataprovider/simpleprovider/TagsProvider.java:
```diff
@@ -0,0 +1,47 @@
+import fr.neamar.kiss.pojo.TagDummyPojo;
+import fr.neamar.kiss.searcher.Searcher;
+
+public class TagsProvider implements IProvider {
```
you can probably extend SimpleProvider here

Neamar commented on app/src/main/java/fr/neamar/kiss/result/TagDummyResult.java:
```diff
@@ -0,0 +1,140 @@
+        paint.getTextBounds(glyph, 0, glyph.length(), b);
+        canvas.drawText(glyph, 0, glyph.length(), width / 2.f - b.centerX(), height / 2.f - b.centerY(), paint);
+
+//        rectF.set(b);
```
dead code?

TBog replied:
@Neamar it is useful for debug or if you like how it looks, maybe add a toggle for it.

Neamar replied:
ok, let's keep it then!

Neamar commented:
Ok! Code does look good, if you can just merge the conflicts and notify me I'll merge and release :)

TBog commented:
I got a lot of conflicts during the merge and now if I remove a favorite from the bar the bar does not get updated :(

EmmanuelMess commented on app/src/main/java/fr/neamar/kiss/result/TagDummyResult.java:
```diff
@@ -0,0 +1,191 @@
+        }
+        if (block == null)
+            drawAsHole = false;
+        else if ("DINGBATS".equals(block.toString()))
```
Why not use an int?

EmmanuelMess commented on app/src/main/java/fr/neamar/kiss/result/TagDummyResult.java:
```diff
@@ -0,0 +1,191 @@
+            block = Character.UnicodeBlock.of(codepoint);
+        } catch (IllegalArgumentException ignored) {
+        }
+        if (block == null)
```
IMO this shouldn't be a gigantic if else, it should set the variable directly.

TBog replied:
directly to what? How can I avoid the `if else` tree?

EmmanuelMess replied:
Maybe something like:

```
drawAsHole = !(block == null || "DINGBATS".equals(block.toString()) || "EMOTICONS".equals(block.toString()) || "MISCELLANEOUS_SYMBOLS".equals(block.toString()) || "MISCELLANEOUS_SYMBOLS_AND_PICTOGRAPHS".equals(block.toString()) || "SUPPLEMENTAL_SYMBOLS_AND_PICTOGRAPHS".equals(block.toString()) || "TRANSPORT_AND_MAP_SYMBOLS".equals(block.toString()));
```

EmmanuelMess commented on app/src/main/java/fr/neamar/kiss/result/Result.java:
```diff
@@ -404,6 +410,7 @@ public long getUniqueId() {
             this.appResultWeakReference = new WeakReference<>(result);
         }
 
+        @SuppressLint("WrongThread")
```
This doesn't seem like a good idea.

TBog replied:
What would you change?

EmmanuelMess replied:
I would remove the annotation, so that warning keeps showing and can be fixed later.

TBog commented:
Fixed the merge problem I had.
@Neamar can you please check my code again and merge? Thanks.

Neamar commented:
Alright let's get this into master :)

Good job, thanks for your work and patience!
